### PR TITLE
Feature/how claims work

### PR DESCRIPTION
**Pull request by MathiasOhrstrom**
Item from design: [Home Tab #Start]

### Screenshots 
image
image
image
image
image

### Checklist

- [x] Functionality is covered by an integration test
- [x] Functionality is accessible in engineering mode

### CODE REVIEW

File: app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePager.kt [hunk 1]
```diff
@@ -59,57 +68,103 @@ abstract class DismissiblePager : DialogFragment() {
...
-                    pagerIndicator.translationX = translation
-                }
-                if (position == count - 1 && offsetPercentage == 0f) {
-                    onDismiss()
-                    dialog?.dismiss()
+            binding.apply {
+                pager.adapter?.count?.let { count ->
+                    when (items.last()) {
+                        is DismissiblePagerModel.SwipeOffScreen->{
+                            if (position == count - 2) {
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePager.kt:118]: I think we can perhaps remove these position checks in our new paradigm with the `DismissiblePagerModel`. Let's investigate that together.

File: app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePagerAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+import android.content.Context
+import androidx.fragment.app.FragmentManager
+import androidx.fragment.app.FragmentPagerAdapter
+
+class DismissiblePagerAdapter(
+    fragmentManager: FragmentManager,
+    private val data: List<DismissiblePagerModel>,
+    private val context: Context
+) : FragmentPagerAdapter(fragmentManager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePagerAdapter.kt:11]: I think this constant might be deprecated, let's check that out

File: app/src/main/java/com/hedvig/app/feature/home/ui/HomeModel.kt [hunk 3]
```diff
@@ -35,6 +35,8 @@ sealed class HomeModel {
 
     object StartClaimContained : HomeModel()
 
+    data class HowClaimsWork(val pager: List<HomeQuery.HowClaimsWork>): HomeModel()
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/home/ui/HomeModel.kt:38]:
```suggestion
    data class HowClaimsWork(val pages: List<HomeQuery.HowClaimsWork>): HomeModel()
```

File: app/src/main/java/com/hedvig/app/feature/loggedin/ui/LoggedInActivity.kt [hunk 4]
```diff
@@ -19,6 +19,7 @@ import com.hedvig.app.HedvigApplication
 import com.hedvig.app.R
 import com.hedvig.app.databinding.ActivityLoggedInBinding
 import com.hedvig.app.feature.claims.ui.ClaimsViewModel
+import com.hedvig.app.feature.dismissiblepager.DismissiblePagerModel
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/loggedin/ui/LoggedInActivity.kt:22]: Are these imports used?

File: app/src/main/java/com/hedvig/app/util/extensions/ContextExtensions.kt [hunk 5]
```diff
@@ -42,6 +42,12 @@ const val SHARED_PREFERENCE_ASKED_FOR_PERMISSION_PREFIX_KEY =
 
 fun Context.compatColor(@ColorRes color: Int) = ContextCompat.getColor(this, color)
 
+fun Context.themeColor(@AttrRes attrRes: Int): Int {
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/util/extensions/ContextExtensions.kt:45]: I think we already have an extension function called `colorAttr` 🙂

**Review by hedvigoscar — CHANGES_REQUESTED** [review]
As there are a lot of changes to be made, let's pair up on fixing them 🙂

File: app/src/main/res/layout/how_claims_work_button.xml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
```
[the remark this answers is not in the document]
Comment (MathiasOhrstrom) [reply app/src/main/res/layout/how_claims_work_button.xml:3]: I used the frameLayout to center the button

[the remark this answers is not in the document]
Comment (hedvigoscar) [reply app/src/main/res/layout/how_claims_work_button.xml:3]: Of course! That makes sense. We can disregard this.

File: app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePageFragment.kt [hunk 7]
```diff
@@ -12,7 +12,7 @@ import com.hedvig.app.R
 import com.hedvig.app.util.svg.buildRequestBuilder
 import kotlinx.android.synthetic.main.fragment_news.*
 
-class DismissablePageFragment : androidx.fragment.app.Fragment() {
+class DismissiblePageFragment : androidx.fragment.app.Fragment() {
```
[the remark this answers is not in the document]
Comment (MathiasOhrstrom) [reply app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePageFragment.kt:15]: What do you mean?

File: app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePager.kt [hunk 8]
```diff
@@ -59,57 +68,103 @@ abstract class DismissiblePager : DialogFragment() {
...
+            }
+
+            titleLabel?.let { tl ->
+                title.text = resources.getString(tl)
+                title.show()
+            }
+
+            pager.adapter = DismissiblePagerAdapter(childFragmentManager, items, requireContext())
+            pagerIndicator.pager = pager
+            proceed.text = if (items.size > 1) {
```
[the remark this answers is not in the document]
Comment (MathiasOhrstrom) [reply app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePager.kt:96]: Do you mean from content-service?

File: app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePager.kt (hunk 1 above)
Comment (MathiasOhrstrom) [reply app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePager.kt:118]: Dont we need these to check if we are on the page next to the `SwipeOffScreen` fragment?

File: app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePageFragment.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePageFragment.kt:15]:
```kotlin
import androidx.fragment.app.Fragment
// ...

class DismissiblePagerFragment : Fragment() {
```

[the remark this answers is not in the document]
Comment (MathiasOhrstrom) [reply app/src/main/java/com/hedvig/app/feature/dismissiblepager/DismissiblePageFragment.kt:15]: Aahh 🤦

**Comment by hedvigoscar** [comment]
I really like the fixes to the bottom sheets and the navigation bar, but we should try to keep those changes in separate PRs in the future as the scope of this PR is starting to grow quite large 😬

**Comment by MathiasOhrstrom** [comment]
> I really like the fixes to the bottom sheets and the navigation bar, but we should try to keep those changes in separate PRs in the future as the scope of this PR is starting to grow quite large 😬

True i should probably have done that in i another PR, however here are the screenshots 
image
image

**Comment by hedvigoscar** [comment]
Looks amazing 🙂

File: app/src/main/java/com/hedvig/app/feature/home/ui/HomeAdapter.kt [hunk 9]
```diff
@@ -310,6 +317,41 @@ class HomeAdapter(
...
+                fragmentManager: FragmentManager,
+                retry: () -> Unit,
+                requestBuilder: RequestBuilder<PictureDrawable>,
+                tracker: HomeTracker
+            ): Any? = with(binding) {
+                if (data !is HowClaimsWork) {
+                    return invalid(data)
+                }
+                val howClaimsWorkData = data.pages.mapIndexed { index, page ->
+                    if (index == data.pages.size - 1) {
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/home/ui/HomeAdapter.kt:334]: Move this if to the row where we decide which button text we send to the `NoTitlePage`

File: app/src/main/java/com/hedvig/app/feature/home/ui/HomeAdapter.kt [hunk 10]
```diff
@@ -310,6 +317,41 @@ class HomeAdapter(
...
+                        DismissiblePagerModel.NoTitlePage(
+                            ThemedIconUrls.from(page.illustration.variants.fragments.iconVariantsFragment),
+                            page.body,
+                            button.context.getString(R.string.claims_explainer_02_button_next)
+                        )
+                    }
+                }
+                button.setHapticClickListener {
+                    HowClaimsWorkDialog.newInstance(howClaimsWorkData)
+                        .show(fragmentManager, WelcomeDialog.TAG)
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/home/ui/HomeAdapter.kt:350]:
```suggestion
                        .show(fragmentManager, HowClaimsWorkDialog.TAG)
```

File: app/src/main/java/com/hedvig/app/feature/whatsnew/WhatsNewDialog.kt [hunk 11]
```diff
@@ -36,13 +47,23 @@ class WhatsNewDialog : DismissiblePager() {
...
                     PAGES,
-                    ArrayList(pages.map { page ->
-                        DismissiblePagerPage(
-                            ThemedIconUrls.from(page.illustration.variants.fragments.iconVariantsFragment),
-                            page.title,
-                            page.paragraph
-                        )
-                    })
+                    ArrayList(pages.mapIndexed { index, page ->
+                        if (index == items.size - 1) {
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/whatsnew/WhatsNewDialog.kt:51]: Same here!

File: app/src/main/java/com/hedvig/app/feature/welcome/WelcomeViewModel.kt [hunk 12]
```diff
@@ -1,18 +1,22 @@
...
 import com.hedvig.app.util.apollo.ThemedIconUrls
 import e
 import kotlinx.coroutines.launch
 
 class WelcomeViewModel(
-    private val welcomeRepository: WelcomeRepository
-) : ViewModel() {
+    private val welcomeRepository: WelcomeRepository,
+    application: Application
+) : AndroidViewModel(application) {
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/welcome/WelcomeViewModel.kt:17]: There's some risk with doing this due to how we work with `Context` and language resolution. I've got a nice strategy for refactoring it, let's do it together right away.

**Review by hedvigoscar — COMMENTED** [review]
Overall very good stuff. Let's pair program the last parts so that we get it over the finish line!

**Review by hedvigoscar — COMMENTED** [review]
Approved pending design QA. Do not merge until that QA is done.

File: app/src/main/res/layout/dismissible_page_no_title_fragment.xml [hunk 13]
```diff
@@ -8,26 +9,28 @@
...
-        android:layout_width="match_parent"
-        android:layout_height="0dp"
-        android:layout_gravity="center_horizontal"
-        android:layout_marginStart="@dimen/base_margin_triple"
-        android:layout_marginEnd="@dimen/base_margin_triple"
-        android:layout_weight="1"
+        android:layout_width="@dimen/dismissible_page_illustration_size"
+        android:layout_height="@dimen/dismissible_page_illustration_size"
         android:contentDescription="@null"
+        android:maxHeight="200dp"
```
Comment (hedvigoscar) [inline app/src/main/res/layout/dismissible_page_no_title_fragment.xml:15]: extract to dimens

Comment (hedvigoscar) [reply app/src/main/res/layout/dismissible_page_no_title_fragment.xml:15]: actually come to think of it, is this necessary? given that we have a fixed width/height.

**Review by hedvigoscar — APPROVED** [review]
Approved after checking comments

File: app/src/main/res/layout/dismissible_page_no_title_fragment.xml (hunk 13 above)
Comment (MathiasOhrstrom) [reply app/src/main/res/layout/dismissible_page_no_title_fragment.xml]: Ture! I forgot to remove it

**Pull request merged**